online-shop/src/app/product-details/product-details.component.ts

```diff
@@ -1,22 +1,34 @@
-import {Component, OnInit} from '@angular/core';
-import {ActivatedRoute} from "@angular/router";
-import {Product} from "../../product";
-import {products} from "../../product";
-
+import { Component, OnInit } from '@angular/core';
+import {ActivatedRoute, Router} from '@angular/router';
+import { Product } from '../../product';
+import {ProductService} from "../product.service";
 
 @Component({
   selector: 'app-product-details',
   templateUrl: './product-details.component.html',
-  styleUrls: ['./product-details.component.scss']
+  styleUrls: ['./product-details.component.scss'],
 })
 export class ProductDetailsComponent implements OnInit {
   id = 0;
-  product!: Product
-constructor(
-  private route: ActivatedRoute,
-) {}
+  product!: Product;
+  constructor(private route: ActivatedRoute, private productService: ProductService) {}
+
+
+
   ngOnInit(): void {
-   this.id = parseInt(this.route.snapshot.paramMap.get('id')!);
-   this.product = products.find(product => product.id === this.id)!;
+    this.id = parseInt(this.route.snapshot.paramMap.get('id')!);
+    this.productService.getProductById(this.id).subscribe(data => {
```

Comment:
when you work with Observables you have to unsubscribe at the end of component life cycle. If you don't memory leaks will appear or depends on the case, unwanted behaviors.
another option in this case, since is just a presentational component, you can have a look here-> https://angular.io/api/common/AsyncPipe